File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case10_watermark/Case02_InOrderStreamEventWatermark.java
package com.yw.flink.example.javacases.case10_watermark;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;

import java.time.Duration;

/**
 * Flink 有序流中设置watermark
 * 案例:读取socket基站日志数据，每隔5s统计每个基站的通话时长
 */
public class Case02_InOrderStreamEventWatermark {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
//        env.setParallelism(1);
        //读取socket数据
        //001,181,182,busy,1000,10
        DataStreamSource<String> sourceDS = env.socketTextStream("node5", 9999);

        //转换数据流为StationLog 对象DataStream
        SingleOutputStreamOperator<StationLog> stationLogDS = sourceDS.map((MapFunction<String, StationLog>) s -> {
            String[] split = s.split(",");
            return new StationLog(split[0], split[1], split[2], split[3], Long.valueOf(split[4]), Long.valueOf(split[5]));
        });

        //设置watermark
        SingleOutputStreamOperator<StationLog> dsWithWatermark = stationLogDS.assignTimestampsAndWatermarks(
                //给有序事件流设置watermark
                WatermarkStrategy.<StationLog>forMonotonousTimestamps()
                        .withTimestampAssigner((SerializableTimestampAssigner<StationLog>) (stationLog, timestamp) -> {
                            //从数据流中选择事件时间列，这里选择的列必须是毫秒
                            return stationLog.callTime;
                        })
                        //当某些并行度超过该空闲时间后，自动推进watermark
                        .withIdleness(Duration.ofSeconds(5))

        );

        //设置窗口，每隔5秒统计每个基站的通话总时长
        dsWithWatermark.keyBy(stationLog -> stationLog.sid)
                //设置滚动窗口，每隔5生成一个
                .window(TumblingEventTimeWindows.of(Time.seconds(5)))
                .sum("duration").print();
        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case10_watermark/Case02_InOrderStreamEventWatermark.scala
package com.yw.flink.example.scalacases.case10_watermark

import com.yw.flink.example.StationLog
import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time

import java.time.Duration

/**
  * Flink 有序流中设置watermark
  * 案例:读取socket基站日志数据，每隔5s统计每个基站的通话时长
  */
object Case02_InOrderStreamEventWatermark {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //设置隐式转换
    import org.apache.flink.streaming.api.scala._

    //读取socket数据 ，格式：001,181,182,busy,1000,10
    val sourceDS: DataStream[String] = env.socketTextStream("node5", 9999)

    //转换成StationLog对象流
    val stationLogDS: DataStream[StationLog] = sourceDS.map(line => {
      val split: Array[String] = line.split(",")
      StationLog(split(0), split(1), split(2), split(3), split(4).toLong, split(5).toLong)
    })

    //设置watermark
    val dsWithWatermark: DataStream[StationLog] = stationLogDS.assignTimestampsAndWatermarks(
      //有序流中设置watermark
      WatermarkStrategy.forMonotonousTimestamps[StationLog]()
        .withTimestampAssigner(new SerializableTimestampAssigner[StationLog] {
          //设置事件时间，必须是毫秒
          override def extractTimestamp(stationLog: StationLog, l: Long): Long = stationLog.callTime
        })
        //设置并行度空闲时间，自动推荐watermark
        .withIdleness(Duration.ofSeconds(5))
    )

    //设置窗口，每隔5秒设置窗口
    dsWithWatermark.keyBy(_.sid)
      .window(TumblingEventTimeWindows.of(Time.seconds(5)))
      .sum("duration")
      .print()
    env.execute()

  }

}
